rust/ql/lib/codeql/rust/controlflow/internal/generated/CfgNodes.qll

```diff
@@ -0,0 +1,3657 @@
+      private BreakExpr node;
+
+      BreakExprCfgNode() { node = this.getAstNode() }
+
+      /** Gets the underlying `BreakExpr`. */
+      BreakExpr getBreakExpr() { result = node }
+
+      /**
+       * Gets the `index`th attr of this break expression (0-based).
+       */
+      Attr getAttr(int index) { result = node.getAttr(index) }
+
+      /**
+       * Gets any of the attrs of this break expression.
+       */
+      Attr getAnAttr() { result = this.getAttr(_) }
+
+      /**
+       * Gets the number of attrs of this break expression.
+       */
+      int getNumberOfAttrs() { result = count(int i | exists(this.getAttr(i))) }
+
+      /**
+       * Gets the expression of this break expression, if it exists.
+       */
+      ExprCfgNode getExpr() {
+        any(ChildMapping mapping).hasCfgChild(node, node.getExpr(), this, result)
+      }
+
+      /**
+       * Holds if `getExpr()` exists.
+       */
+      predicate hasExpr() { exists(this.getExpr()) }
+
+      /**
+       * Gets the lifetime of this break expression, if it exists.
+       */
+      Lifetime getLifetime() { result = node.getLifetime() }
+
+      /**
+       * Holds if `getLifetime()` exists.
+       */
+      predicate hasLifetime() { exists(this.getLifetime()) }
+    }
+
+    final private class ParentCallExpr extends ParentAstNode, CallExpr {
+      override predicate relevantChild(AstNode child) {
+        none()
+        or
+        child = this.getExpr()
+      }
+    }
+
+    /**
+     * A function call expression. For example:
+     * ```rust
+     * foo(42);
+     * foo::<u32, u64>(42);
+     * foo[0](42);
+     * foo(1) = 4;
+     * ```
+     */
+    final class CallExprCfgNode extends CfgNodeFinal, CallExprBaseCfgNode {
+      private CallExpr node;
+
+      CallExprCfgNode() { node = this.getAstNode() }
+
+      /** Gets the underlying `CallExpr`. */
+      CallExpr getCallExpr() { result = node }
+
+      /**
+       * Gets the expression of this call expression, if it exists.
+       */
+      ExprCfgNode getExpr() {
+        any(ChildMapping mapping).hasCfgChild(node, node.getExpr(), this, result)
+      }
+
+      /**
+       * Holds if `getExpr()` exists.
+       */
+      predicate hasExpr() { exists(this.getExpr()) }
+    }
+
+    final private class ParentCallExprBase extends ParentAstNode, CallExprBase {
+      override predicate relevantChild(AstNode child) { none() }
+    }
+
+    /**
+     * A function or method call expression. See `CallExpr` and `MethodCallExpr` for further details.
+     */
+    final class CallExprBaseCfgNode extends CfgNodeFinal, ExprCfgNode {
+      private CallExprBase node;
+
+      CallExprBaseCfgNode() { node = this.getAstNode() }
+
+      /** Gets the underlying `CallExprBase`. */
+      CallExprBase getCallExprBase() { result = node }
+
+      /**
+       * Gets the argument list of this call expression base, if it exists.
+       */
+      ArgList getArgList() { result = node.getArgList() }
```

Comment:
If one has a `CallExprBaseCfgNode` and do `c.getArgList().getArg(n)` one gets a plain `Expr`. That _could_ be a footgun.

Reply:
It could, but I don't think there is an easy/nice way to get rid of this auto-generated predicate. The best solution, I think, is to get rid of the intermediate `ArgList` node altogether in the extractor.